src/Exception/RestException.php

```diff
@@ -0,0 +1,109 @@
+<?php
+
+namespace Www\Exception;
+
+use YapepBase\Exception\Exception;
+
+class RestException extends Exception
+{
+
+    const CODE_PARAMETER_ERROR = 'ParameterError';
+    const CODE_UNAUTHENTICATED = 'UnauthenticatedError';
+    const CODE_PAYMENT_REQUIRED = 'PaymentRequired';
+    const CODE_UNAUTHORIZED = 'UnauthorizedError';
+    const CODE_METHOD_NOT_SUPPORTED = 'MethodNotSupported';
+    const CODE_INTERNAL_ERROR = 'InternalError';
+
+    const MSG_PARAMETER_ERROR = 'An invalid parameter was sent to the endpoint, or a required parameter is missing';
+    const MSG_UNAUTHENTICATED = 'This endpoint requires authentication but no "Authorization" header is sent, or the token is invalid';
+    const MSG_PAYMENT_REQUIRED = 'The authenticated account is in debt';
+    const MSG_UNAUTHORIZED = 'The authenticated account is not authorized to perform the requested action';
+    const MSG_METHOD_NOT_SUPPORTED = 'The requested endpoint exists, but does not support the specified method';
+    const MSG_INTERNAL_ERROR = 'An internal error occured while serving the request. The error has been logged, please try your request again later';
+
+    const PARAM_ERROR_MISSING = 'missing';
+    const PARAM_ERROR_DUPLICATE = 'duplicate';
+    const PARAM_ERROR_INVALID = 'invalid';
+    const PARAM_ERROR_MUST_BE_EMPTY = 'mustBeEmpty';
+
+    /** @var string  */
+    protected $errorCode;
+
+    /** @var array */
+    protected $params = [];
+
+    public function __construct($errorCode, $message = null, array $params = [], $code = 0, \Exception $previous = null, $data = null)
+    {
+        $this->errorCode = $errorCode;
+        $this->params    = $params;
+
+        $message = null === $message ? $this->getDefaultMessageForErrorCode() : $message;
```

Comment:
It may make sense, especially considering the string 0 would be considered empty as well. I'm hoping we'd be returning better messages then '' or '0', but you never know if there'd be a use case for it.